File: app/src/main/java/org/watsi/uhp/fragments/EnrollmentMemberPhotoFragment.java
package org.watsi.uhp.fragments;

import android.app.AlertDialog;
import android.content.DialogInterface;
import android.view.View;
import android.widget.Button;
import android.widget.Toast;

import org.watsi.domain.entities.IdentificationEvent;
import org.watsi.domain.entities.Member;
import org.watsi.domain.entities.Photo;
import org.watsi.domain.repositories.MemberRepository;
import org.watsi.uhp.R;
import org.watsi.uhp.managers.LegacyNavigationManager;
import org.watsi.uhp.presenters.EnrollmentPresenter;

import java.io.IOException;

import javax.inject.Inject;

public class EnrollmentMemberPhotoFragment extends PhotoFragment<Member> {

    private IdentificationEvent mIdEvent;
    private EnrollmentPresenter enrollmentPresenter;

    @Inject MemberRepository memberRepository;

    @Override
    int getTitleLabelId() {
        return R.string.enrollment_member_photo_fragment_label;
    }

    @Override
    int getFragmentLayoutId() {
        return R.layout.fragment_capture_photo;
    }

    @Override
    public boolean isFirstStep() {
        return true;
    }

    @Override
    public void nextStep() {
        if (!mSyncableModel.shouldCaptureFingerprint()) {
            AlertDialog.Builder builder = new AlertDialog.Builder(getContext());
            builder.setMessage(R.string.enrollment_fingerprint_confirm_completion);
            builder.setPositiveButton(android.R.string.yes, new DialogInterface.OnClickListener() {
                @Override
                public void onClick(DialogInterface dialog, int which) {
                    memberRepository.save(mSyncableModel);
                    getNavigationManager().setMemberDetailFragment(mSyncableModel, mIdEvent);
                    enrollmentPresenter.confirmationToast().show();
                }
            });
            builder.setNegativeButton(android.R.string.no, new DialogInterface.OnClickListener() {
                @Override
                public void onClick(DialogInterface dialog, int which) {
                    dialog.dismiss();
                }
            });
            builder.show();
        } else if (mSyncableModel.shouldCaptureNationalIdPhoto()) {
            getNavigationManager().setEnrollmentIdPhotoFragment(mSyncableModel, mIdEvent);
        } else {
            getNavigationManager().setEnrollmentContactInfoFragment(mSyncableModel, mIdEvent);
        }
    }

    @Override
    void handleSetupFailure() {
        getNavigationManager().setCurrentPatientsFragment();
        Toast.makeText(getContext(), R.string.generic_error_message, Toast.LENGTH_LONG).show();
    }

    @Override
    void additionalSetup(View view) {
        ((Button) view.findViewById(R.id.photo_btn)).setText(R.string.enrollment_member_photo_btn);
        mIdEvent = (IdentificationEvent) getArguments().getSerializable(LegacyNavigationManager.IDENTIFICATION_EVENT_BUNDLE_FIELD);

        if (!mSyncableModel.shouldCaptureFingerprint()) {
            mSaveBtn.setText(R.string.enrollment_complete_btn);
        }

        enrollmentPresenter = new EnrollmentPresenter(mSyncableModel, getContext());
    }

    @Override
    void onPhotoCaptured(Photo photo) throws IOException {
        mSyncableModel.setLocalMemberPhoto(photo);
        mSaveBtn.setEnabled(true);
    }
}

File: app/src/main/java/org/watsi/uhp/fragments/EnrollmentMemberPhotoFragment.kt
package org.watsi.uhp.fragments

import android.app.AlertDialog
import android.content.Intent
import android.graphics.BitmapFactory
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import dagger.android.support.DaggerFragment
import io.reactivex.android.schedulers.AndroidSchedulers
import kotlinx.android.synthetic.main.fragment_capture_photo.photo
import kotlinx.android.synthetic.main.fragment_capture_photo.photo_btn
import kotlinx.android.synthetic.main.fragment_capture_photo.save_button
import org.threeten.bp.Clock
import org.watsi.device.managers.Logger

import org.watsi.domain.entities.Member
import org.watsi.domain.repositories.IdentificationEventRepository
import org.watsi.domain.repositories.MemberRepository
import org.watsi.domain.repositories.PhotoRepository
import org.watsi.uhp.R
import org.watsi.uhp.activities.SavePhotoActivity
import org.watsi.uhp.managers.NavigationManager

import java.util.UUID

import javax.inject.Inject

class EnrollmentMemberPhotoFragment : DaggerFragment() {

    @Inject lateinit var clock: Clock
    @Inject lateinit var navigationManager: NavigationManager
    @Inject lateinit var memberRepository: MemberRepository
    @Inject lateinit var identificationEventRepository: IdentificationEventRepository
    @Inject lateinit var photoRepository: PhotoRepository
    @Inject lateinit var logger: Logger

    private var photoIds: Pair<UUID, UUID>? = null

    companion object {
        const val CAPTURE_PHOTO_INTENT = 1
        const val PARAM_MEMBER = "member"

        fun forMember(member: Member): EnrollmentMemberPhotoFragment {
            val fragment = EnrollmentMemberPhotoFragment()
            fragment.arguments = Bundle().apply {
                putSerializable(PARAM_MEMBER, member)
            }
            return fragment
        }
    }

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        activity.setTitle(R.string.enrollment_member_photo_fragment_label)
        return inflater?.inflate(R.layout.fragment_capture_photo, container, false)
    }

    override fun onViewCreated(view: View?, savedInstanceState: Bundle?) {

        photo_btn.setOnClickListener {
            startActivityForResult(Intent(activity, SavePhotoActivity::class.java), CAPTURE_PHOTO_INTENT)
        }

        save_button.setOnClickListener {
            val member = arguments.getSerializable(PARAM_MEMBER) as Member

            val updatedMember = member.copy(photoId = photoIds?.first,
                                            thumbnailPhotoId = photoIds?.second)
            if (!member.requiresFingerprint(clock)) {
                val builder = AlertDialog.Builder(context)
                builder.setMessage(R.string.enrollment_fingerprint_confirm_completion)
                builder.setPositiveButton(android.R.string.yes) { _, _ ->
                    memberRepository.save(updatedMember)

                    identificationEventRepository.openCheckIn(member.id).subscribe({
                        navigationManager.goTo(CurrentMemberDetailFragment.forIdentificationEvent(it))
                    }, {
                        // TODO: handle error
                    }, {
                        navigationManager.goTo(CheckInMemberDetailFragment.forMember(updatedMember))
                    })

                    Toast.makeText(activity, "Enrollment completed", Toast.LENGTH_LONG).show()
                }
                builder.setNegativeButton(android.R.string.no) { dialog, _ -> dialog.dismiss() }
                builder.show()
            } else {
                navigationManager.goTo(EnrollmentFingerprintFragment.forMember(updatedMember))
            }
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        val (photoIds, error) = SavePhotoActivity.parseResult(resultCode, data, logger)
        photoIds?.let {
            this.photoIds = it
            photoRepository.find(it.second).observeOn(AndroidSchedulers.mainThread())
                    .subscribe({ thumbnailPhoto ->
                thumbnailPhoto.bytes?.let { photoBytes ->
                    val thumbnailBitmap = BitmapFactory.decodeByteArray(photoBytes, 0, photoBytes.size)
                    photo.setImageBitmap(thumbnailBitmap)
                }
            }, {
                // TODO: handle error
            })
        }
        error?.let {
            // TODO: handle error
        }
    }
}
